### ISSUE TITLE
DNS leaks can't be solved.

### ISSUE THREAD

**Issue by Ergus** [jamesmcm/vopono#89]
Hi:

I am using vopono to connect to protonvpn and I have tried everything but https://www.dnsleaktest.com/ shows my isp in spite of protonvpn has protection.

--dns option does not change the dns 

and adding `block-outside-dns` to the .ovpn hangs when trying to connect.

**Comment by jamesmcm** [comment]
Can you check the content of `/etc/netns/NAMESPACENAME/resolv.conf` when running?

I remember having issues with the ProtonVPN DNS when writing the sync code - https://github.com/jamesmcm/vopono/blob/master/src/providers/protonvpn/openvpn.rs#L39

**Comment by Ergus** [comment]
Hi:

The file content seems to be fine at least the values seems to be set, but I am not sure if they are fine...

With vopono I get:

![image](https://user-images.githubusercontent.com/6975176/120117944-22e7cc80-c190-11eb-8a86-a3211440f945.png)

But when using network Manager I get:

![image](https://user-images.githubusercontent.com/6975176/120118128-4d865500-c191-11eb-89c2-9582108f8f3f.png)

In this case I use the same server (COnfig nl-free-09.protonvpn.com.tcp.ovpn) but I get very different inets and subnets... Is this intended??

**Comment by jamesmcm** [comment]
That inet stuff is fine, and it seems alright using the internal DNS server (i.e. that will only be accessible within the tunnel).

Just to check if you run bash within the namespace and then run `dig google.com` for example - what DNS servers does it use?

**Comment by Ergus** [comment]
Interesting, bash seems to show the right dns server with dig... But dnsleaktest.com from firefox (from the namespace of course) still shows the one from my ISP. It is using the vpn because the ip is one from proton, not my public one but the Standard test shows the same than without the namespace. 

OTOH I still tried to add `block-outside-dns` to the config and vopono hangs during startup.... So we are missing something here ;(.

**Comment by jamesmcm** [comment]
Apparently `block-outside-dns` should only affect Windows clients anyway. It might be blocking on missing credentials?

In any case, can you try with Chromium or another browser? In case it's due to Firefox doing DNS over HTTPS or some other setting?

**Comment by Ergus** [comment]
> Apparently `block-outside-dns` should only affect Windows clients anyway. It might be blocking on missing credentials?
> 
But in any case it shouldn't block vopono right?

> In any case, can you try with Chromium or another browser?

I just tried Chromium and midori with exactly the same result right ip wrong dns. 

>In case it's due to Firefox doing DNS over HTTPS or some other setting?


I actually disabled DNS over HTTPS on firefox; same result. And also tried with private windows (just in case)

**Comment by jamesmcm** [comment]
I just tested this with the Japan free OpenVPN server and it worked fine for me with `google-chrome-stable`:

![Japanese DNS](https://i.imgur.com/SC8LNdI.png)

**Comment by Ergus** [comment]
> I just tested this with the Japan free OpenVPN server and it worked fine for me with `google-chrome-stable`:
> 

Hi is this from proton? or someone else?

If it is a proton one. Did you do anything  special? Because I can't solve this. I have the same issue with any browser I use except lyx and elinks. So maybe there is something wrong in my environment?? Something that should be documented maybe?

**Comment by jamesmcm** [comment]
Yeah, I just used `vopono sync`, chose the free one and logged in. Then ran `google-chrome-stable` with vopono.

Maybe try to check if anything is overriding the DNS somehow?

**Comment by Ergus** [comment]
> Yeah, I just used `vopono sync`, chose the free one and logged in. Then ran `google-chrome-stable` with vopono.
> 
> Maybe try to check if anything is overriding the DNS somehow?

I just tried with google-chrome-stable and I have the same result. I am starting to think that maybe this is related with networkmanager?

Are you using networkmanager??

**Comment by Ergus** [comment]
I found another hint. I rebooted the system to an old debian installation and there vopono seems to work very fine and there are not dns leaks....

The debian installation uses the same environment I have (i3wm+networkmanager) so I maybe this has nothing to do with network manager. The only difference is the os (Debian vs Arch). So probably there is something in arch that makes the gui applications to bypass the terminal environment during dns search for requests...

**Comment by jamesmcm** [comment]
It's strange, I use NetworkManager myself, and have [openresolv](https://wiki.archlinux.org/title/Openresolv) installed but it's not an issue at all.

I don't see anything weird on the Wiki page either - https://wiki.archlinux.org/title/Domain_name_resolution

**Comment by Ergus** [comment]
I  had `systemd-resolvconf` but I just switched to `openresolv`, restarted and the issue is still there... I ran out of ideas

**Comment by Ergus** [comment]
BTW In firefox I tried `about:networking#dns` but there is not info about how firefox searches for a DNS either

**Comment by billop** [comment]
I'm experiencing the same issue with mullvad.

**Comment by fosspill** [comment]
I'm having this issue as well. Both with ungoogled-chromium and firefox.
If I use the config files and connect with wg myself it works fine, however.

**Comment by niki-on-github** [comment]
Maybe it has something to do with the `nsswitch.conf` setting. See [dns-leaks-with-network-namespaces](https://philipdeljanov.com/posts/2019/05/31/dns-leaks-with-network-namespaces/). Since vopono does not generate a custom `nsswitch.conf` config the default file will be used. Depending on the configuration of the distributions this could maybe cause such problems.

**Comment by jamesmcm** [comment]
Thanks, I can try to test it by switching to systemd-resolved.

**Comment by jamesmcm** [comment]
If you're using systemd-resolved, it'd be great if you could try vopono 0.9.0 now thanks to PR #144 from @niki-on-github

**Comment by rendeko** [comment]
My apologies for the necrobump but I wanted to confirm that issue still persists specifically on NixOS 26.05 with vopono 0.10.18. The best solution I could find after a lot of debugging was similar to @masaeedu's suggestion, wrapping the command being tunneled in firejail and refusing it access to the nscd socket. Prefixing the process to be tunneled with `firejail --noprofile --blacklist=/var/run/nscd/socket` was sufficient to stop DNS leaks.
For example:
`vopono exec --custom <configFile> "firejail --noprofile --blacklist=/var/run/nscd/socket chromium"`

I run Vopono through Bash aliases that handle most of the arguments, so I can just prefix the firejail wrapper in those, but it is still a silent DNS leak unless the user catches it or finds this old issue. Is there any way to fix this upstream on distros where running nscd is mandatory so something like firejail isn't required?
